feat: make contract and hook types work without a generated output

sverps opened the pull request:
Fixes #254

A dev that doesn't need any custom contract can simply remove the `hardhat_contracts.ts` import in the scaffold.config, and set `contracts: null`.

Typescript won't complain due to types resolving differently when `contracts` is set to `null` in the config.

carletex commented on packages/nextjs/scaffold.config.ts:
```diff
@@ -1,9 +1,23 @@
+import { Abi } from "abitype";
 import * as chains from "wagmi/chains";
+import contracts from "~~/generated/hardhat_contracts";
```
This still makes the app crash if you do `yarn start` and the file is not there.

How about:

```ts
let contracts;
try {
  contracts = require("~~/generated/hardhat_contracts");
} catch (error) {
  // Silently fail. The builder might not want to use a contract deployed from SE2
}
```

sverps replied:
I'm guessing that will break the narrow typing. I don't think there's a way to get around that. Next best thing we can do is explain very well in the setup that if the user wants a contract, he has to deploy it first, and if he doesn't want it, he should modify this in the config.

carletex replied:
> The aim of my PR is to make the changes a dev needs to make to use se-2 without contract as minimal as possible.

I have the same aim :)

What about?

```ts
// If you don't use custom contract for this SE-2 app:
// - Remove the import line below
// - Se "contracts: null" in the scaffoldConfig below
import contracts from "~~/generated/hardhat_contracts";

```

sverps replied:
Ah yes :+1: 

Would probably also be a good idea to update the readme a bit and explain how to use the scaffold.config

carletex replied:
db5d16bd28bdde35176faa7eb9da6dabef3549be

carletex replied:
I leave updating the README for another PR. We also need to include some other stuff.

carletex commented on packages/nextjs/scaffold.config.ts:
Not introduced in this PR, but since we are dealing with the `hardhat_contract` file maybe we can take a look (happy to create another issue)

![image](https://user-images.githubusercontent.com/2486142/227493366-d6385388-af7c-4e32-8565-a12a9dc785e9.png)

"Sometimes" (Haven't been able to reproduce consistently, but happen on app start up.. sometimes) this show.

It's weird because it's referencing the **JSON** file. Maybe because it gets created and the removed? Or because it has the same name as the **TS** one so the imports get confused?

sverps replied:
Hmmm, interesting. I think your guess is correct. When you deploy an update to your contract (and next is already running), then there's a short time where `hardhad_contacts.json` exists, and it tries importing that, but then it gets removed... etc.

I think we should fix it by not putting the `json` output in the next folder.
`"deploy": "hardhat deploy --export-all ../nextjs/generated/hardhat_contracts.json && yarn generateTsAbis",`

We could export the `json` file to `/packages/hardhat/generated` or something, and then use that as the source for the `generateTsAbis`. Makes a bit more sense that way.

carletex replied:
> We could export the json file to /packages/hardhat/generated or something, and then use that as the source for the generateTsAbis. Makes a bit more sense that way

Yep! This would work :)

Thanks

carletex commented on packages/nextjs/scaffold.config.ts:
```diff
@@ -19,6 +33,9 @@ const scaffoldConfig = {
   alchemyApiKey: process.env.NEXT_PUBLIC_ALCHEMY_API_KEY || "oKxs-03sij-U_N0iOlrSsZFr29-IqbuF",
+
+  // This is the contract data that was generated by "yarn deploy", or null if no local contracts are used
+  contracts,
```
When setting this to `null`, I get the following type error in the project:

![image](https://user-images.githubusercontent.com/2486142/227499236-e42fc606-4d0c-4368-b0ac-4ba1c7d24ac6.png)

sverps replied:
Hmmm, strange that this didn't fail before. I do see it now, though. Fixed it :+1:

carletex commented on packages/nextjs/hooks/scaffold-eth/useAutoConnect.ts:
@sverps Added this check (for the burnerwallet autoconnect), and another one in the Faucet, to avoid UI & console errors when no contracts.

carletex commented on packages/hardhat/scripts/generateTsAbis.ts:
```diff
@@ -1,21 +1,22 @@
 function main() {
-  const GENERATED_PATH = "../nextjs/generated/hardhat_contracts";
+  const TARGET_DIR = "../nextjs/generated/";
 
   const fileContent = Object.entries(allGeneratedContracts).reduce((content, [chainId, chainConfig]) => {
     return `${content}${parseInt(chainId).toFixed(0)}:${JSON.stringify(chainConfig, null, 2)},`;
   }, "");
 
+  if (!fs.existsSync(TARGET_DIR)) {
+    fs.mkdirSync(TARGET_DIR);
+  }
   fs.writeFileSync(
-    `${GENERATED_PATH}.ts`,
+    `${TARGET_DIR}hardhat_contracts.ts`,
     prettier.format(`export default {${fileContent}} as const;`, { parser: "typescript" }),
   );
```
Do we want to keep the `json` file?

It might be confusing: sometimes I search for "hardhat_contract" in my IDE to check that file and now it shows both.

If we remove it, we wouldn't need to create a folder, It could be a tmp file inside the scripts folder that gets removed.

Or am I missing something?

sverps replied:
Fair point. I figured there was no harm in keeping it and being able to see the actual output, but your point makes sense. I'll remove it. Might even make more sense to just write it in an actual temp dir. I'll fix it :+1:

sverps commented:
Found another confusing thing: the `generateTsAbi` script just spits out an error when called directly (which is logical since the json doesn't exist). This can be rather confusing for a user that is exploring which scripts exist in the `package.json`. He might not understand that this script is only used as part of the deploy and is not to be used directly (even though the name suggests it could).

So I removed the separate script for it and just added it to the deploy script instead.

carletex commented:
> So I removed the separate script for it and just added it to the deploy script instead.

Yep, nice one. Make sense!

This looks great. Merging it. 

Thank you @sverps
